drivers: digipot: add generic interface and AD528x driver support

jocelynmass opened the pull request:
This patch introduces a generic digital potentiometer (digipot) driver interface in Zephyr along with an implementation for the Analog Devices AD528x series.

Key changes:
	•	Added a digipot driver API to provide a unified interface for controlling digital potentiometers.
	•	Implemented support for AD5280 and AD5282 devices, including RDAC read/write operations over I2C.
	•	Added Kconfig and device tree bindings to configure and instantiate AD528x devices.
	•	Provides basic initialization, wiper position set/get functions, and integration with Zephyr’s device model.

This enables Zephyr applications to use digital potentiometers in a standardized way, making it easier to add support for additional digipot devices in the future.

Tested on:
	•	STM32 platform using I2C devices configured via device tree.

jocelynmass committed:
drivers: add digital potentiometers driver interface

This commit introduces a generic driver interface for digital
potentiometers. The API provides standard operations for setting and
reading the wiper position on devices that support absolute positioning

Signed-off-by: Jocelyn Masserot <jocemass@gmail.com>

jocelynmass committed:
driver: add AD528x digipot driver

Adds a driver for the Analog Devices AD528x family of digital
potentiometers. These devices support absolute wiper positioning and
wiper readback using i2c.

Signed-off-by: Jocelyn Masserot <jocemass@gmail.com>

carlescufi commented:
Hi @jocelynmass. Thanks for your contribution. Since this introduces a new API, it needs to go through the Architecture WG meeting. Agendas are emailed to the [devel@ mailing list](https://lists.zephyrproject.org/g/devel/messages), and additional details can be found here:
https://github.com/zephyrproject-rtos/zephyr/wiki/Architecture-Working-Group

henrikbrixandersen commented on include/zephyr/drivers/digipot.h:
```diff
@@ -0,0 +1,134 @@
+ * @retval 0 if successful
+ * @retval < 0 if setting wiper failed
+ */
+static inline int digipot_get_wiper(const struct device *dev, uint8_t channel, uint16_t *pos)
```
Same comment as above re: stepping vs. absolute positions.

henrikbrixandersen reviewed (commented):
Perhaps we don't need to include `_wiper` in all of these API function names? What other operations do you foresee on a digital potentiometer that do not operate on the wiper?

henrikbrixandersen reviewed (changes requested):
See comments above.

carlescufi commented:
> Thanks for the heads up. Unfortunately, I have a conflict at that time today. I’ll watch the recording.

@jocelynmass unfortunately we did not have the time to cover this topic today, so I will schedule it for next Tuesday.

jocelynmass commented on include/zephyr/drivers/digipot.h:
```diff
@@ -0,0 +1,134 @@
+ * @retval 0 if successful
+ * @retval < 0 if setting wiper failed
+ */
+static inline int digipot_set_wiper(const struct device *dev, uint8_t channel, uint16_t pos)
```
Thanks for the comment. 
Most commonly used digital potentiometers (AD5282, MCP41xxx, MCP4251, AD525x, DS3502, etc.) support absolute wiper positioning and readback, so I chose an API that exposes the absolute set/get operations directly.
You're right that a subset of devices only provide relative increment/decrement commands. These devices could still be supported by resetting the wiper at init and maintaining an internal software position, but that might not be ideal for all drivers.
If we later need to support purely step-based digipots, we can extend the API with optional `step_up()` / step_down()` operations, or add a capability flag.
For now, the absolute API matches the majority of hardware and keeps the common use case clean.

jocelynmass replied:
My initial focus was on serial-controlled digipots (i2c/spi), which are also the only type integrated with linux. 
If the intent is for this API to cover both absolute-position and step-only devices, I can add optional step_up() / step_down() functions.

jocelynmass commented on include/zephyr/drivers/digipot.h:
```diff
@@ -0,0 +1,134 @@
+ * @retval 0 if successful
+ * @retval < 0 if setting wiper failed
+ */
+static inline int digipot_shutdown(const struct device *dev, uint8_t channel)
```
Good question, you're right that "shutdown" is not universally supported across all digital potentiometers.
Some digipot families (e.g. Microchip MCP41xxx/42xxx and a few ADI parts) expose a hardware shutdown or Hi-Z mode, with the goal of reducing the power consumption.
I included the shutdown API because several commonly used devices support it, but I'm fine making it optional or capability-based if that makes the API more generic. We could add a feature flag (e.g. DIGIPOT_CAP_SHUTDOWN)
or remove shutdown entirely if it's considered too device-specific.
Let me know what direction you'd prefer, I can update the API accordingly.

jocelynmass replied:
You’re right, I’m still getting familiar with Zephyr and its overall architecture. After reviewing the documentation, this approach makes complete sense. I’ll remove the shutdown API from the digipot driver for now and plan to add proper power management support in a follow-up PR.

matt-wood-ct commented:
Have you had any thoughts on my request for NV digipot coverage?
I'd be keen to see some way to store and recall the NV value separate from the volatile value

jocelynmass commented:
> MCP41U83

My apologies for not addressing your question earlier. I chose to wait for the outcome of the architecture meeting before adding further API functions, to ensure that the design evolves consistently with the broader feedback.

jocelynmass commented:
> Also, @jocelynmass could you sign-off-by your commits please? See [our commit message guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines) to find out how to format your commit messages, and at [our contribution workflow](https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow) to understand how to update your Pull Request.

Unfortunately, I’m unable to attend the meeting as I have recurring conflicts during the Architecture call times.
I have updated the branch accordingly and added signed-off-by tags to all commits.

mbolivar reviewed (changes requested):
Please squash your fixup commit as needed

henrikbrixandersen commented:
> Got it, I will change this later today

@jocelynmass Will you be returning to this?

jocelynmass commented:
> > Got it, I will change this later today
> 
> @jocelynmass Will you be returning to this?

Sorry for the delay, I got sidetracked with other stuff. I’m planning to finish the changes before the end of this week.

ttmut commented on drivers/digipot/Kconfig.ad528x:
```diff
@@ -0,0 +1,13 @@
+	depends on DT_HAS_ADI_AD5280_ENABLED \
+		|| DT_HAS_ADI_AD5282_ENABLED
+	help
+	  Enable the driver for the Analog Devices AD528x digipot
```
```suggestion
	  Enable the driver for the Analog Devices AD528x digipot

```
Missing newline?

ttmut commented on drivers/digipot/digipot_ad528x.c:
```diff
@@ -0,0 +1,155 @@
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/digipot.h>
+#include <zephyr/logging/log.h>
```
```suggestion
```
I do not see any logging functions being used.

ttmut commented on drivers/digipot/digipot_ad528x.c:
```diff
@@ -0,0 +1,155 @@
+	DEVICE_DT_INST_DEFINE(index, ad528x_init, NULL, NULL,				\
+							&config_ad5282_##index,						\
+							POST_KERNEL, CONFIG_DIGIPOT_INIT_PRIORITY,	\
+							&ad528x_driver_api);
```
```suggestion
/* Helper that instantiates a driver for AD5280 (1 RDAC) */
#define INST_AD5280(index)												\
	static const struct ad528x_config config_ad5280_##index = {			\
		.bus = I2C_DT_SPEC_INST_GET(index),								\
		.rdac_nb = 1,													\
	};																	\
	DEVICE_DT_INST_DEFINE(index, ad528x_init, NULL, NULL, 				\
							&config_ad5280_##index,						\
							POST_KERNEL, CONFIG_DIGIPOT_INIT_PRIORITY,	\
							&ad528x_driver_api);

/* Helper that instantiates a driver for AD5282 (2 RDACs) */
#define INST_AD5282(index)												\
	static const struct ad528x_config config_ad5282_##index = {			\
		.bus = I2C_DT_SPEC_INST_GET(index),								\
		.rdac_nb = 2,													\
	};																	\
	DEVICE_DT_INST_DEFINE(index, ad528x_init, NULL, NULL,				\
							&config_ad5282_##index,						\
							POST_KERNEL, CONFIG_DIGIPOT_INIT_PRIORITY,	\
							&ad528x_driver_api);
```
The escape characters seem to have stretched a bit too far.

ttmut commented on include/zephyr/drivers/digipot.h:
```diff
@@ -0,0 +1,109 @@
+static inline int digipot_reset(const struct device *dev, uint8_t channel)
+{
+    const struct digipot_driver_api *api = (const struct digipot_driver_api *)dev->api;
+    return api->reset_position(dev, channel);
```
```suggestion
	return DEVICE_API_GET(digipot, dev)->reset_position(dev, channel);
```

ttmut commented on drivers/digipot/digipot_ad528x.c:
```diff
@@ -0,0 +1,155 @@
+	const struct ad528x_config *config = dev->config;
+	uint8_t cmd = 0;
+
+	if (channel > (config->rdac_nb-1)) {
```
```suggestion
	if (channel > (config->rdac_nb - 1)) {
```

ttmut commented on drivers/digipot/digipot_ad528x.c:
```diff
@@ -0,0 +1,155 @@
+	const struct ad528x_config *config = dev->config;
+	uint8_t cmd = 0;
+
+	if (channel > (config->rdac_nb-1)) {
```
```suggestion
	if (channel > (config->rdac_nb - 1)) {
```

ttmut reviewed (changes requested):
Adding a few comments. Thanks!

mbolivar reviewed (approved):
+1 for the DT bindings only, did not review the rest.
